cloudsdk/src/test/java/io/particle/android/sdk/cloud/ParticleCloudTest.kt

```diff
@@ -0,0 +1,224 @@
+import io.particle.android.sdk.cloud.Responses.CardOnFileResponse
+import io.particle.android.sdk.cloud.Responses.ClaimCodeResponse
+import io.particle.android.sdk.cloud.Responses.DeviceMeshMembership
+import io.particle.android.sdk.cloud.Responses.FirmwareUpdateInfoResponse
+import io.particle.android.sdk.cloud.Responses.LogInResponse
+import io.particle.android.sdk.cloud.Responses.MeshNetworkRegistrationResponse
+import io.particle.android.sdk.cloud.Responses.Models.CompleteDevice
+import io.particle.android.sdk.cloud.Responses.PingResponse
+import io.particle.android.sdk.cloud.Responses.ReadDoubleVariableResponse
+import io.particle.android.sdk.cloud.Responses.ReadIntVariableResponse
+import io.particle.android.sdk.cloud.Responses.ReadObjectVariableResponse
+import io.particle.android.sdk.cloud.Responses.ReadStringVariableResponse
+import io.particle.android.sdk.cloud.Responses.SimpleResponse
+import io.particle.android.sdk.cloud.models.*
+import io.particle.android.sdk.persistance.AppDataStorage
+import io.particle.android.sdk.persistance.SensitiveDataStorage
+import io.particle.android.sdk.utils.Broadcaster
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import retrofit.RestAdapter.LogLevel
+import retrofit.client.Response
+import retrofit.mime.TypedOutput
+import java.util.*
+import java.util.concurrent.Executors
+
+
+private const val FAKE_API_PORT = 8080
+private val FAKE_API_URL = HttpUrl.parse("http://localhost:$FAKE_API_PORT")
+
+
+class ParticleCloudTest {
+
+    lateinit var cloud: ParticleCloud
+    lateinit var mockServer: MockWebServer
+
+    @Before
+    fun setUp() {
+        initCloud()
+    }
+
+    @After
+    fun teardown() {
+        // close the socket, stop the listening thread
+        mockServer.shutdown()
+    }
+
+    private fun initCloud() {
+        // start the mock webserver
+        mockServer = MockWebServer()
+        mockServer.start(FAKE_API_PORT)
+
+        // build up a few dependency fakes
+        SDKGlobals.appDataStorage = FakeAppDataStorage()
+        SDKGlobals.sensitiveDataStorage = FakeSensitiveDataStorage()
+
+        val tokenDelegate = TokenGetterDelegate { "d34db33fd34db33fd34db33fd34db33fd34db33f" }
+        val credsProvider = object : OauthBasicAuthCredentialsProvider {
+            override fun getClientSecret(): String = "I_AM_A_CLIENT_SECRET"
+            override fun getClientId(): String = "I_AM_A_CLIENT_ID"
+        }
+
+        // use actual ApiFactory to build other dependencies
+        val factory = ApiFactory(FAKE_API_URL, LogLevel.FULL, tokenDelegate, credsProvider)
+
+        // build the cloud itself
+        cloud = ParticleCloud(
+            FAKE_API_URL,
+            factory.buildNewCloudApi(),
+            factory.buildNewIdentityApi(),
+            SDKGlobals.appDataStorage!!,
+            FakeBroadcaster(),
+            Gson(),
+            Executors.newSingleThreadExecutor()
+        )
+    }
```

Comment:
i could see you eventually wanting to split up cloud-related tests into different files as you expand coverage - e.g. `ParticleCloudEventsTest.kt`, `ParticleCloudFunctionTest.kt`, etc. 

longer-term it'd be nice to extract this setup / mocking stuff such that actual tests are more prominent and those fakes can be reused across the test suite.

Reply:
Yeah, agreed on all points!